File: examples/observability-cascade/observability-cascade-java-node/src/main/java/org/organicprogramming/observability/cascade/javaholon/internal/RelayServer.java
package org.organicprogramming.observability.cascade.javaholon.internal;

import gen.describe_generated;
import io.grpc.stub.StreamObserver;
import org.organicprogramming.holons.Describe;
import org.organicprogramming.holons.Serve;
import org.organicprogramming.observability.cascade.javaholon.api.PublicApi;
import relay.v1.Relay;
import relay.v1.RelayServiceGrpc;

import java.util.List;

public final class RelayServer extends RelayServiceGrpc.RelayServiceImplBase {
    @Override
    public void tick(Relay.TickRequest request, StreamObserver<Relay.TickResponse> responseObserver) {
        responseObserver.onNext(PublicApi.tick(request));
        responseObserver.onCompleted();
    }

    public static void listenAndServe(String listenUri, boolean reflect, List<Serve.MemberRef> members) throws Exception {
        Describe.useStaticResponse(describe_generated.StaticDescribeResponse());
        Serve.runWithOptions(
                normalizeListenUri(listenUri),
                List.of(new RelayServer()),
                new Serve.Options()
                        .withReflect(reflect)
                        .withSlug("observability-cascade-node-java")
                        .withMemberEndpoints(members));
    }

    private static String normalizeListenUri(String listenUri) {
        if (listenUri != null && listenUri.startsWith("tcp://:")) {
            return "tcp://0.0.0.0:" + listenUri.substring("tcp://:".length());
        }
        return listenUri;
    }
}

File: examples/observability-cascade/observability-cascade-kotlin-node/src/main/kotlin/org/organicprogramming/observability/cascade/kotlinholon/internal/RelayServer.kt
package org.organicprogramming.observability.cascade.kotlinholon.internal

import org.organicprogramming.holons.Serve
import org.organicprogramming.observability.cascade.kotlinholon.api.PublicApi
import relay.v1.Relay
import relay.v1.RelayServiceGrpcKt

class RelayServer : RelayServiceGrpcKt.RelayServiceCoroutineImplBase() {
    override suspend fun tick(request: Relay.TickRequest): Relay.TickResponse =
        PublicApi.tick(request)

    companion object {
        fun listenAndServe(listenUri: String, reflect: Boolean, members: List<Serve.MemberRef>) {
            Serve.runWithOptions(
                normalizeListenUri(listenUri),
                listOf(RelayServer()),
                Serve.Options(
                    reflect = reflect,
                    slug = "observability-cascade-node-kotlin",
                    memberEndpoints = members,
                ),
            )
        }

        private fun normalizeListenUri(listenUri: String): String =
            if (listenUri.startsWith("tcp://:")) {
                "tcp://0.0.0.0:${listenUri.removePrefix("tcp://:")}"
            } else {
                listenUri
            }
    }
}
